File: algorithms/java/HashTable.java
// HashTable.java
// By Lizhuo You
// This is a data structure that stores values with index. List.
// The index depends on the size of the hashtable and the value's hash.
// In lots of situations the value can be accessed with O(1) efficiency.
import java.util.ArrayList;
import java.util.LinkedList;

public class HashTable<T> {
    private  ArrayList<LinkedList<T>> table;
    private int size;       // it accounts the num of the values.
    private int tableSize;      // it is the full size of the table.
    private double maxLoad;     // max load percentage
    public HashTable(){
        size = 0;
        tableSize = 10;
        maxLoad = 0.6;
        clear(10);
    }

    private void clear(int n){
        // it clears the table
        table = new ArrayList<LinkedList<T>>();
        for(int i = 0;i<n;i++){
            table.add(null);        // add spots for the table.
        }
    }

    public void add(T val){
        // add a value to the table.
        int pos = Math.abs(val.hashCode())%table.size();
        LinkedList<T>list = table.get(pos);

        if(list == null){           // ensure there is a LinkedList at the index to hold the value
            list = new LinkedList<T>();
            table.set(pos,list);
        }

        list.add(val);
        size++;

        if ((double) size / (double) tableSize > maxLoad) {     // check if the load amount is over max percent
            resize();
        }
    }

    public void resize() {
        // resize the table to increase its spots which allows more values to be entered.
        ArrayList<LinkedList<T>> tmp = table;
        size = 0;
        tableSize = table.size()*10;        // update the full size of the table
        clear(table.size() * 10);       // clear and increase the size of the table to hold more values.
        for (LinkedList<T> lst : tmp) {
            if(lst != null){
                for (T val : lst) {
                    add(val);       // add back all the values to the table
                }
            }
        }
    }

    public void remove(T val) {
        // it removes a val from the table
        if (contains(val)) {        // first ensure the value is in the table
            int pos = Math.abs(val.hashCode())%table.size();        // get position of the value
            LinkedList<T>list = table.get(pos);         // get the list in which the value is put

            if (list.size() == 1) {     // remove the value.
                table.set(pos, null);
            } else {
                list.remove(val);
            }

            size--;
        }
    }

    public boolean contains(T val) {
        // it checks if the val is in the table and returns true/false
        int pos = Math.abs(val.hashCode())%table.size();
        if (table.get(pos) == null) {
            return false;       // if there is nothing at the pos, the val is not in the table.
        }

        for (T n : table.get(pos)) {
            if (n.hashCode() == val.hashCode()) {       // check if the val actually exists.
                return true;
            }
        }

        return false;
    }

    public void setMaxLoad(double percent) {
        // set the max load of the table
        if (percent >= 0.1 && percent <= 0.8) {     // check if the new maxLoad is valid
            maxLoad = percent;
        }
    }

    public void setLoad(double percent) {
        // set the max load of the table and resize the table if the max load decreased
        if (percent >= 0.1 && percent <= 0.8) {     // check if the new maxLoad is valid
            if (percent < maxLoad) {
                if((double) size / (double) tableSize > percent) {  // check if it is necessary to resize the table.
                    resize();       // reset the hash table to force the load to be what was asked for
                }
            }
            maxLoad = percent;
        }
    }

    public double getLoad() {
        // return how full the hash table is
        return (double) size / (double) tableSize;
    }

    public ArrayList<T> toArray() {
        // return an ArrayList of the stored objects
        ArrayList<T> ans = new ArrayList<T>();      // answer
        for (LinkedList<T> lst : table) {
            if (lst != null) {
                for(T val: lst){
                    ans.add(val);   // store the val to the answer
                }
            }
        }

        return ans;     // return the answer.
    }

    @Override
    public String toString() {
        // returns a string with the contents of the table.
        String ans = "";
        for (LinkedList<T> lst : table) {
            if (lst != null) {
                for(T val: lst){
                    ans += " ," + val;
                }
            }
        }
        if (!ans.equals("")) {
            ans = ans.substring(2);
        }
        return ans;
    }
}

File: structures/hashTable.cpp
#include <iostream>
#include <string>
#include <list>

using namespace std;

class hashTable{
private:
    int size, tableSize;
    double maxLoad;
    list<int> *table{};

    void clear(int n){
        // it clears and resize the table
        delete [] table;
        table = new list<int>[n];
    }

    void resize() {
        // resize the table to increase its spots which allows more values to be entered.
        auto *lst = new list<int>;
        for(int i = 0; i < tableSize; i++){
            list<int> l = table[i];
            while(!l.empty()){
                lst->push_back(l.front());
                l.pop_front();
            }
        }

        size = 0;
        tableSize *= 2;        // update the full size of the table
        clear(tableSize );       // clear and increase the size of the table to hold more values.
        while(!lst->empty()){
            add(lst->front());
            lst->pop_front();
        }

        delete lst;
    }

public:
    hashTable(){
        size = 0;
        tableSize = 10;
        maxLoad = 0.6;
        clear(10);
    }

    int hashFunction(int val){
        if(val < 0) val *= -1;
        return val % tableSize;
    }

    void add(int val){
        // add a value to the table.
        int pos = hashFunction(val);
        list<int> *lst = &table[pos];
        lst->push_back(val);
        size++;
        if (getLoad() > maxLoad) {     // check if the load amount is over max percent
            resize();
        }
    }

    double getLoad(){
        return (double) size / (double) tableSize;
    }

    void setMaxLoad(double percent){
        // set the max load of the table, wich is restricted from 0.2 to 0.8
        if (percent >= 0.1 && percent <= 0.8) {     // check if the new maxLoad is valid
            while(getLoad()>maxLoad) resize();
            maxLoad = percent;
        }
    }

    bool contains(int val){
        int index = hashFunction(val);
        list<int>::iterator it;
        for (it = table[index].begin(); it != table[index].end(); ++it) {
            if(val == *it) return true;
        }
        return false;
    }

    void remove(int val){
        int index = hashFunction(val);
        list<int>::iterator it;
        for (it = table[index].begin(); it != table[index].end(); ++it) {
            if(val == *it) {
                table[index].remove(val);
                size--;
                return;
            }
        }
    }

    string toString(){
        string ans = "[";
        for(int i = 0; i < tableSize; i++){
            list<int> lst = table[i];
            while(!lst.empty()){
                ans += to_string(lst.front()) + ", ";
                lst.pop_front();
            }
        }
        return ans.substr(0, ans.length()-2)+"]";
    }
};

int main(){
    hashTable hashT = hashTable();
    hashT.add(0);
    hashT.add(11);
    hashT.add(13);
    hashT.add(14);
    hashT.add(12);
    hashT.add(15);
    cout << hashT.toString() << endl;
    hashT.add(10);
    hashT.add(1);
    hashT.add(3);
    hashT.add(4);
    hashT.add(2);
    hashT.add(7);
    cout << hashT.toString() << endl;
    if(hashT.contains(7)) cout << "It contains " << 7 << endl;
    hashT.remove(7);
    hashT.setMaxLoad(0.2);
    cout << hashT.toString() << endl;
    return 0;
}